keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,776 @@
+    # Default for nodes not matching any other rule
+    NodeSelector:    nil,
+    # For existing API this is the same as NodeCIDRMaskSize
+    PerNodeMaskSize: 24,
+},
+{
+    IPFamily:        IPV4,
+    IPCIDRBlock:     "172.16.0.0/14",
+    # Another range, also allocate-able to any node
+    NodeSelector:    nil,
+    PerNodeMaskSize: 24,
+},
+{
+    IPFamily:        IPV4,
+    IPCIDRBlock:     "10.0.0.0/8",
+    NodeSelector:    { key: "np" op: "IN" value:["np1"] },
+    PerNodeMaskSize: 26,
+},
+{
+    IPFamily:        IPV4,
+    IPCIDRBlock:     "192.168.0.0/16",
+    NodeSelector:    { key: "np" op: "IN" value:["np2"] },
+    PerNodeMaskSize: 26,
+},
+{
+    IPFamily:        IPV4,
+    IPCIDRBlock:     "5.2.0.0/16",
+    NodeSelector:    { "np": "np3" },
+    PerNodeMaskSize: 20,
+},
+{
+    IPFamily:        IPV6,
+    IPCIDRBlock:     "fd12:3456:789a:1::/64",
+    NodeSelector:    { "np": "np3" },
+    PerNodeMaskSize: 112,
+},
+...
+]
+```
+
+Given the above config, a valid potential configuration might be:
+
+```
+{“np”: “np1”} --> "10.0.0.0/26"
+{“np”: “np2”} --> "192.16.0.0/26"
+{“np”: “np3”} --> "5.2.0.0/20", "fd12:3456:789a:1::/112"
+{“np”: “np4”} --> "172.16.0.0/24"
+```
+
+### Controller
+
+Implement a new [NodeIPAM controller](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/nodeipam) The controller will set up watchers on the `ClusterCIDRConfig` objects and the `Node` objects.
+
+#### Data Structures
+We want to use a set of radix trees to store allocation. This is a commonly used pattern (linux uses it to store routing tables). This will track the ranges that are already allocated. We will also store the range allocated to each node. This is to handle conflicts in case some entity modifies the podCIDR value of the Node object.
+
+#### Startup
+- Fetch the values of the cluster-cidr and node-cidr-mask-size arguments passed to kube-controller-manager.
+- Fetch list of `ClusterCIDRConfig` and build internal data structure (radix trees)
+- Fetch list of `Node`s. Check each node for `PodCIDRs`
+    - If `PodCIDR` is set, mark the allocation in the internal data structure and store this association with the node.
+    - If `PodCIDR` is set, but is not part of one of the tracked `ClusterCIDRConfig`s, emit a K8s event but do nothing.
+    - If `PodCIDR` is not set, save Node for allocation in the next step.
+After processing all nodes, allocate ranges to any nodes without Pod CIDR(s) [Same logic as Node Added event]
+
+#### Event Watching Loops
+
+##### Node Added
+Go through the list of `ClusterCIDRConfig`s and find ranges matching the node selector from each family. Attempt to allocate Pod CIDR(s) with the given per-node size. If that `ClusterCIDRConfig` cannot fit a node, search for another `ClusterCIDRConfig`.
+
+If no `ClusterCIDRConfig` matches the node, or if all matching `ClusterCIDRConfig`s are full, raise a K8s event and put the Node back on the queue for allocation (infinite retries).
```

Comment:
what is doing current controller? is doing the same?

Reply:
Yes. The range allocator keeps trying and does not bring up the node.